lib/auth/userloginstate/generator.go

```diff
@@ -364,3 +431,49 @@ func (g *Generator) LoginHook(ulsService services.UserLoginStates) func(context.
 		return trace.Wrap(err)
 	}
 }
+
+// validateRoles is a helper function which validates that a given list of roles exist.
+func (g *Generator) validateRoles(ctx context.Context, roles []string) ([]string, []error, error) {
+	var missingRoles []string
+	var notFoundErrs []error
+
+	for _, role := range roles {
+		_, err := g.access.GetRole(ctx, role)
+		if err != nil {
+			if trace.IsNotFound(err) {
+				missingRoles = append(missingRoles, role)
+				notFoundErrs = append(notFoundErrs, err)
+				continue
+			}
+			return nil, nil, trace.Wrap(err)
+		}
+	}
+
+	if len(missingRoles) > 0 {
+		return missingRoles, notFoundErrs, nil
+	}
+
+	return nil, nil, nil
+}
+
+// emitSkippedAccessListEvent emits an audit log event to warn that an access list was skipped due to it referencing a non-existent role.
```

Comment:
Maybe for the godoc here we just say invalid and don't hint at the reason. Today the only reason is that it may reference a non-existent role, but we may add more ways to invalidate access lists in the future and I imagine we would re-use the same audit event.

```suggestion
// emitSkippedAccessListEvent emits an audit log event to indicate that an invalid
// access list could not be applied during user login.
```